horovod/common/ops/adasum_cuda_ring.cc

```diff
@@ -0,0 +1,332 @@
+// Copyright 2019 Microsoft. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "adasum_cuda_ring.h"
+
+namespace horovod {
+namespace common {
+
+void Ring::InitRing(int tmp[], bool _isFat, int rank, int size) {
+  load = 0;
+  isFat = _isFat;
+  for (int i = 0; i < 8; i++)
+    loop[i] = tmp[i];
+
+  for (int j = 0; j < size; j++) { // go through allranks
+    if (rank == loop[j]) {
+      prevGPU = loop[(j-1+size) % size];
+      nextGPU = loop[(j+1+size) % size];
+    }
+  }
+}
+
+int Ring::GetAfterLoad(int message_len) {
+  if (!isFat)
+    return 2*(load+message_len);
+  else
+    return (load+message_len);
+}
+
+void Ring::AddLoad(int message_len) {
+  load += message_len;
+}
+
+void Ring::ReduceLoad(int message_len) {
+  load -= message_len;
+}
+
+Message::Message(MPIContext* mpi_context)
+  : mpi_context(mpi_context) {
+}
+
+void Message::InitMessage(Ring* _ring, int _rank, int _ring_starter_rank, int _count, void* _grad_buf, void* _recv_buf, DataType _datatype, MPI_Comm _comm, int _tag) {
+  comm = _comm;
+  count = _count;
+  tag = _tag;
+  ring = _ring;
+  rank = _rank;
+  ring_starter_rank = _ring_starter_rank;
+  leg = 0;
+  grad_buf = _grad_buf;
+  recv_buf = _recv_buf;
+  datatype = _datatype;
+  cuda_stream_sync = false;
+  Start();
+}
+
+ReduceMessage::ReduceMessage(MPIContext* mpi_context)
+  : Message(mpi_context) {
+}
+
+void ReduceMessage::Start() {
+
+  auto mpi_datatype = mpi_context->GetMPIDataType(datatype);
+  if (rank == ring_starter_rank) {
+    MPI_Isend(grad_buf, count, mpi_datatype, ring->nextGPU, tag, comm, &req);
+  } else {
+    MPI_Irecv(recv_buf, count, mpi_datatype, ring->prevGPU, tag, comm, &req);
+  }
+}
+
+bool ReduceMessage::Test() {
+  auto mpi_datatype = mpi_context->GetMPIDataType(datatype);
+
+  int flag = 0;
+  if (leg == 2)
+    return true;
+  if (!cuda_stream_sync) {
+    MPI_Test(&req, &flag, MPI_STATUS_IGNORE);
+  }
+  if (flag == 1 || cuda_stream_sync) {
+    leg++;
+    if (leg == 2) {
+      ring->ReduceLoad(count);
+      return true;
+    }
+    if (leg == 1) {
+      static cudaStream_t adasum_stream;
```

Comment:
which scope do you intend to use this stream? i.e. message, ring, allrings. It seems like this one stream is shared by all messages in all rings. If so, can we move this to Allrings or whichever scope you want to use this for?

Reply:
I tried using multiple streams previously, but based on layerids instead of rings. Since the kernel computation is very cheap and most of the overhead comes from the dispatching host thread being blocked by the launch of the new kernel if it's not sent to a separate stream, I think using one stream is fine. But I will try to quickly experiment with using different streams for different rings just to check.